libraries/hotstuff/chain_pacemaker.cpp

```diff
@@ -112,9 +112,31 @@ namespace eosio { namespace hotstuff {
       return _chain->is_builtin_activated( builtin_protocol_feature_t::instant_finality );
    }
 
-   void chain_pacemaker::get_state( finalizer_state& fs ) const {
-      if (enabled())
-         _qc_chain.get_state( fs ); // get_state() takes scare of finer-grained synchronization internally
+   void chain_pacemaker::get_state(finalizer_state& fs) const {
+      if (! enabled())
+         return;
+
+      // lock-free state version check
+      uint64_t current_state_version = _qc_chain.get_state_version();
+      if (_state_cache_version != current_state_version) {
+         finalizer_state current_state;
+         {
+            csc prof("stat");
+            std::lock_guard g( _hotstuff_global_mutex ); // lock IF engine to read state
+            prof.core_in();
+            _qc_chain.get_state(current_state);
+            current_state_version = _qc_chain.get_state_version(); // get potentially fresher version
```

Comment:
This is a little awkward. Is there a reason why we don't include the version in `struct finalizer_state`?

Reply:
The version is only relevant for the cache. It's not related to anything outside thread synchronization.

Reply:
OK, resolving this for now.